### DISCUSSION TITLE
link_only fails with large df (~350 mio rows, 16 cols)

### DISCUSSION THREAD

**Discussion by fritze512** [moj-analytical-services/splink#792] [General]
### What happens?

I have df_1 with 350 mio. rows where dedupe_only works fine. If I try to link_only df1 with df_2 (10 mio. rows) i get "Error: Invalid Error: std::bad_alloc". link_only works if I reduce df_1 to around 0.2 (= 70 mio. rows).

Here the full error:

Traceback (most recent call last):

  File "/home/wdalvai/.local/lib/python3.9/site-packages/spyder_kernels/py3compat.py", line 356, in compat_exec
    exec(code, globals, locals)

  File "/mnt/it/IT/Ent/DAM/KI/Deduplizierung_Baseline_Modell/programm_ausfuehren.py", line 201, in <module>
    clusters = linker.cluster_pairwise_predictions_at_threshold(predictions, 0)

  File "/home/wdalvai/.local/lib/python3.9/site-packages/splink/linker.py", line 1256, in cluster_pairwise_predictions_at_threshold
    cc = solve_connected_components(

  File "/home/wdalvai/.local/lib/python3.9/site-packages/splink/connected_components.py", line 463, in solve_connected_components
    representatives = linker._execute_sql_pipeline()

  File "/home/wdalvai/.local/lib/python3.9/site-packages/splink/linker.py", line 345, in _execute_sql_pipeline
    dataframe = self._sql_to_splink_dataframe_checking_cache(

  File "/home/wdalvai/.local/lib/python3.9/site-packages/splink/linker.py", line 418, in _sql_to_splink_dataframe_checking_cache
    splink_dataframe = self._execute_sql_against_backend(

  File "/home/wdalvai/.local/lib/python3.9/site-packages/splink/duckdb/duckdb_linker.py", line 186, in _execute_sql_against_backend
    self._con.execute(sql).fetch_df()

Error: Invalid Error: std::bad_alloc

### To Reproduce

from splink.duckdb.duckdb_linker import DuckDBLinker
import splink.duckdb.duckdb_comparison_library as cl

settings = {
    "link_type": "link_only",
    "blocking_rules_to_generate_predictions": blocker_full,
    "comparisons": [cl.exact_match("GESCHLECHT")],
    "retain_matching_columns": False,
    "retain_intermediate_calculation_columns": False
    }

input_aliases = ["old", "new"]

linker = DuckDBLinker([df_1, df_2], settings, input_table_aliases=input_aliases)

# Comparison pairs

# Create comparison pairs and cluster_ids
predictions = linker.predict()
df_predictions = predictions.as_pandas_dataframe(limit=50)

clusters = linker.cluster_pairwise_predictions_at_threshold(predictions, 0)
df_clusters = clusters.as_pandas_dataframe(limit=50)

### OS:

RockyLinux 8.5

### Splink version:

3.3.0

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by RobinL** [comment]
You may be trying to create too many comparisons.  To check, you can try the following to see how many comparisons your blocking rules will create:

```
linker.cumulative_num_comparisons_from_blocking_rules_chart()
```

There's an example [here](https://moj-analytical-services.github.io/splink/demos/example_deduplicate_50k_synthetic.html)

There's more information on scaling jobs for big data here:
https://moj-analytical-services.github.io/splink/topic_guides/backends.html
https://moj-analytical-services.github.io/splink/topic_guides/drivers_of_performance.html



I'm going to move this to a discussion for now, because I think the error is arising due to the size of your data and using DuckDB, as opposed to a bug in Splink
